Add a Micrometer extension

ebullient opened the pull request:
Introduce micrometer extension to provide support for micrometer metrics. Acts as an alternative to the smallrye metrics extension. Each extension provides a MetricsCapabilityBuildItem, which is a SimpleBuildItem; a project's build will fail with a message if both are enabled.

ebullient committed:
Micrometer extension

Enable MP Metrics if API is present by default

ebullient committed:
hibernate-orm + micrometer

ebullient committed:
jaeger + micrometer

ebullient committed:
mongodb + micrometer

kenfinnigan commented on integration-tests/micrometer-jmx/pom.xml:
```diff
@@ -0,0 +1,144 @@
+      <version>${project.version}</version>
+    </dependency>
+
+    <dependency>
```
Is `micrometer-core` needed as a dependency if the micrometer extension is present?

ebullient replied:
I removed this yesterday. ;)

kenfinnigan commented on integration-tests/micrometer-jmx/src/main/resources/application.properties:
```diff
@@ -0,0 +1,11 @@
+quarkus.log.category."io.netty".level=INFO
+quarkus.log.category."org.apache".level=INFO
+
+quarkus.micrometer.export.prometheus.enabled=false
```
We need to explicitly disable the prometheus export when the dependency isn't present?

ebullient replied:
We shouldn't. I may have just to be sure.. this might also be left over from how things were structured in my other project. Good observation.

ebullient replied:
fixed.

kenfinnigan commented on integration-tests/micrometer-mp-metrics/pom.xml:
```diff
@@ -0,0 +1,149 @@
+      <version>${project.version}</version>
+    </dependency>
+
+    <dependency>
```
Again, is `micrometer-core` needed?

kenfinnigan commented on integration-tests/micrometer-mp-metrics/src/main/resources/application.properties:
```diff
@@ -0,0 +1,13 @@
+quarkus.log.category."org.apache".level=INFO
+quarkus.log.category."org.jboss.resteasy".level=INFO
+
+quarkus.micrometer.binder.mp-metrics.enabled=true
```
Can the presence of the API be enough to enable it, or we need to be specific?

ebullient replied:
This one needs to be specific because (for the foreseeable future), the smallrye API will be around. I could leave it as default on if the classes are found and just .. let that be a coping mechanism...

ebullient replied:
I've adjusted this extension so that MP Metrics compat layer is on if MP Metrics API is present.

kenfinnigan commented on extensions/micrometer/runtime/src/main/resources/META-INF/quarkus-extension.yaml:
```diff
@@ -0,0 +1,11 @@
+  - "metrics"
+  - "metric"
+  - "prometheus"
+  guide: "https://github.com/ebullient/quarkus-micrometer-extension"
```
Can't remember if we'd discussed or not, did you want to include a guide with this PR?

ebullient replied:
Yes. I can and should.

ebullient replied:
Latest commit has the beginnings of a guide. We should talk about what else should be in there..

kenfinnigan reviewed (commented):
Overall it looks great, just a few minor comments

ebullient committed:
Micrometer guide

kenfinnigan commented on docs/src/main/asciidoc/micrometer-metrics.adoc:
```diff
@@ -0,0 +1,283 @@
+----
+
+This command generates a Maven project, that imports the `micrometer` extension and the micrometer
+prometheus registry as dependencies.
```
Does this add the prometheus registry dependency automatically? I thought it would just be the micrometer extension only and the prometheus registry needs to be manually added?

kenfinnigan replied:
Not sure if you sure this comment @ebullient?

kenfinnigan commented:
Other than the one comment on the guide, I think it's good from my side.

Next week we should move it out of draft status to have a chance of making 1.8 release

gsmet reviewed (changes requested):
I haven't looked at the code in detail but spotted a few issues.

Could you take care of them?

gsmet commented:
CR1 will be released on September 2nd so if you plan to get this one, it needs to be merged by September 1st.

ebullient commented on extensions/micrometer/runtime/src/main/resources/META-INF/quarkus-extension.yaml:
```diff
@@ -0,0 +1,11 @@
+name: "Micrometer - SLF4J for metrics"
```
One of their phrases. It is like SLF4J (the logging API that adapts to other things), but for metrics.

gsmet replied:
I think it's confusing and we should remove that. It's marketing stuff, it really has no value here.

ebullient replied:
I did.

ebullient commented:
> Could you take care of them?

Building now. Thanks, @gsmet! =)

gsmet commented:
It will also need a rebase.

ebullient commented:
> It will also need a rebase.

Naturally.

The documentation step failed with this: 

```
2020-08-28T04:53:31.1411300Z [ERROR] Failed to execute goal on project quarkus-integration-test-mongodb-panache: Could not resolve dependencies for project io.quarkus:quarkus-integration-test-mongodb-panache:jar:999-SNAPSHOT: The following artifacts could not be resolved: io.quarkus:quarkus-micrometer:jar:999-SNAPSHOT, io.quarkus:quarkus-micrometer-deployment:pom:999-SNAPSHOT: Could not find artifact io.quarkus:quarkus-micrometer:jar:999-SNAPSHOT -> [Help 1]
2020-08-28T04:53:31.1411694Z [ERROR] 
2020-08-28T04:53:31.1412083Z [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
2020-08-28T04:53:31.1412453Z [ERROR] Re-run Maven using the -X switch to enable full debug logging.
2020-08-28T04:53:31.1412709Z [ERROR] 
2020-08-28T04:53:31.1412879Z [ERROR] For more information about the errors and possible solutions, please read the following articles:
2020-08-28T04:53:31.1413194Z [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
2020-08-28T04:53:31.1413444Z [ERROR] 
2020-08-28T04:53:31.1413602Z [ERROR] After correcting the problems, you can resume the build with the command
2020-08-28T04:53:31.1414018Z [ERROR]   mvn <args> -rf :quarkus-integration-test-mongodb-panache
2020-08-28T04:53:33.0350879Z ##[error]Process completed with exit code 1.
```

I am pretty sure the dependencies for the integration test are correct, but I wonder if the Documentation CI step is missing something?

gsmet commented:
Question: is there any native tests? I don't see any addition of a module to `ci-actions.yml` so if there are any, they aren't run on CI.

ebullient commented:
> Question: is there any native tests? I don't see any addition of a module to `ci-actions.yml` so if there are any, they aren't run on CI.

Right. There are native tests (nothing fancy) that I was running in my own repo.. I forgot about the bucket definitions in ci-actions.yml. Should I add them to one of the existing categories? Make a new one?

gsmet commented:
I would add them together with metrics and add a 5 minutes more to the timeout. 

Also better rebase to get the latest CI fixes.

ebullient commented:
> I would add them together with metrics and add a 5 minutes more to the timeout.
> 
> Also better rebase to get the latest CI fixes.

Rebased. smallrye-metrics doesn't have dedicated integration tests (micrometer does to account for different backends). I added two of them to one of the misc buckets. Can change it if you want.

ebullient commented:
There are two classloading problems in the non-micrometer-integration-tests running in native mode that were touched to use the micrometer extension instead of mp-metrics. I'm working on it.

gsmet commented:
There are issues in the doc:
```
2020-08-31T03:11:25.2708106Z [INFO] asciidoctor: ERROR: ed/config/quarkus-micrometer.adoc: line 173: table missing leading separator; recovering automatically
2020-08-31T03:11:25.2801064Z [INFO] asciidoctor: ERROR: ed/config/quarkus-micrometer.adoc: line 228: table missing leading separator; recovering automatically
2020-08-31T03:11:25.3504994Z [INFO] Converted /home/runner/work/quarkus/quarkus/docs/src/main/asciidoc/micrometer-metrics.adoc
2020-08-31T03:11:25.3591888Z [ERROR] asciidoctor: ERROR: ed/config/quarkus-micrometer.adoc: line 173: table missing leading separator; recovering automatically
2020-08-31T03:11:25.3603054Z [ERROR] asciidoctor: ERROR: ed/config/quarkus-micrometer.adoc: line 228: table missing leading separator; recovering automatically
```

Please make a visual pass locally on the rendered output of the documentation when you're adding a significant piece of doc.

ebullient commented:
> There are issues in the doc:
> 
> ```
> 2020-08-31T03:11:25.2708106Z [INFO] asciidoctor: ERROR: ed/config/quarkus-micrometer.adoc: line 173: table missing leading separator; recovering automatically
> 2020-08-31T03:11:25.2801064Z [INFO] asciidoctor: ERROR: ed/config/quarkus-micrometer.adoc: line 228: table missing leading separator; recovering automatically
> 2020-08-31T03:11:25.3504994Z [INFO] Converted /home/runner/work/quarkus/quarkus/docs/src/main/asciidoc/micrometer-metrics.adoc
> 2020-08-31T03:11:25.3591888Z [ERROR] asciidoctor: ERROR: ed/config/quarkus-micrometer.adoc: line 173: table missing leading separator; recovering automatically
> 2020-08-31T03:11:25.3603054Z [ERROR] asciidoctor: ERROR: ed/config/quarkus-micrometer.adoc: line 228: table missing leading separator; recovering automatically
> ```
> 
> Please make a visual pass locally on the rendered output of the documentation when you're adding a significant piece of doc.

Yep. I saw that. I'm trying to sort that out. I have something locally (I needed to sleep ;) )

kenfinnigan reviewed (approved):
Minor comment on the doc related to what dependencies get added

ebullient commented:
All set, I think. =)

jmartisk commented on .github/workflows/ci-actions.yml:
```diff
@@ -501,11 +501,13 @@ jobs:
               picocli-native
               gradle
+              micrometer-mp-metrics
+              micrometer-prometheus
```
Do we have `micrometer-jmx` somewhere in the CI? Isn't it missing?

gsmet replied:
Does it have native tests? If not, there's no need.

jmartisk replied:
Oh. You're right.

ebullient replied:
I didn't set up all of them. There is a native test for Micrometer + JMX, but I didn't want to extend the build time too much. If/when we have incremental builds, I would make a separate bucket for micrometer and go to town.

gsmet commented:
OK, let's get this one merged. Thanks @ebullient !
